### PR TITLE
Map 104 / 108 - Actual GetFeature Request

### PR DESCRIPTION
**Pull request by wchen329**
In this pull request:

* GetFeature and GetCapabilities is actually implemented most of the way! This, however, means that launching a GetFeature request on /thredds/wfs/ and/or a non-existing dataset no longer works... This is because GetFeature (and GetCapabilities) now determine Feature Types based on the specified dataset (and will fetch and write geospatial data from those datasets)!

* So a couple of things will require implementing in the future those being: proper bounding boxes (have to zoom in a lot right now), and handling edge cases with TYPENAMES/TYPENAME.

* Line and Point do test data do not currently render in QGIS as of now. DescribeFeatureType must be made to change according to dataset as well in order for this to work.

* The HRU Soil Moist Dataset _will_ render in QGIS, provided the URL includes the correct dataset.

* The WFS Geometry Element is named "catchments_geometry_container"- this could probably change to just "geometry" instead.

### CODE REVIEW

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCSBuilder.java [hunk 1]
```diff
@@ -64,10 +66,13 @@ public static String describe(NetcdfDataset ds, CoordinateSystem cs, Formatter e
...
   private List<Dimension> dims;
+  private List<String> geometrySeriesVarNames;
+  private List<String> geometryContainerNames;
   private SimpleGeometryReader geometryReader;
+  private Map<String, List<String>> geometryContainersAssoc;
   private ProjectionImpl orgProj;
-
-  SimpleGeometryCSBuilder(NetcdfDataset ds, CoordinateSystem cs, Formatter errlog) {
+  
+  public SimpleGeometryCSBuilder(NetcdfDataset ds, CoordinateSystem cs, Formatter errlog) {
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCSBuilder.java:75]: Thanks for adding functionality to keep track of the geometry containers, the geometry variables, and their associations! The usages in WFSFeatureWriter prove that this is invaluable information for producing valid WFS responses.

This class is stretching the limits of the Builder pattern, but so is the DtCoverageCSBuilder that it is based on, so it's not a big deal. I added a JIRA for following up later if we want to take a crack at it.

File: tds/src/main/java/thredds/server/wfs/GMLFeatureWriter.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ private String writePolygon(Polygon poly) {
 
         Polygon polygon = poly;
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/GMLFeatureWriter.java:77]: Could this commented-out code be removed?

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 3]
```diff
@@ -87,32 +96,71 @@ private void describeFeatureType(PrintWriter out, HttpServletRequest hsreq) {
...
-		ArrayList<SimpleGeometry> geometries = new ArrayList<SimpleGeometry>();
-
-//		geometries.add(new CFPoint(50.0, 50.0, null, null, null));
-
-		//ArrayList<Point> points = new ArrayList<Point>();
-		//points.add(new CFPoint(50.0, 50.0, null, null, null));
-		//points.add(new CFPoint(70.0, 50.0, null, null, null));
-		//geometries.add(new CFLine(points));
-
+	private void getFeature(PrintWriter out, HttpServletRequest hsreq, SimpleGeometryCSBuilder sgcs, String ftName) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:100]: Yay! Good bye to hardcoding!

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 4]
```diff
@@ -229,6 +279,25 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
 			String version = null;
 			String service = null;
 			String typeNames = null;
+			String datasetReqPath = null;
+			String actualPath = null;
+			String actualFTName = null;
+			NetcdfDataset dataset = null;
+			
+			if(hsreq.getServletPath().length() > 4) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:287]: Why is this check helpful? Could documentation be added to clarify? Could the numbers here be made into one or two constants whose name illuminates their purpose?

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 5]
```diff
@@ -247,12 +316,17 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
 					service = hsreq.getParameter(paramName);
 				}
 				
-				if(paramName.equalsIgnoreCase("TYPENAMES")) {
+				if(paramName.equalsIgnoreCase("TYPENAMES") || paramName.equalsIgnoreCase("TYPENAME")) {
 					typeNames = hsreq.getParameter(paramName);
+					
+					// Remove namespace header for getFeature
+					if(typeNames != null) if(typeNames.length() > TDSNAMESPACE.length()) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:323]: Thank you for null-checking! Why are these logical expressions in separate `if` statements? Could these `if`s be combined into one `if` with an `&&` ? Due to [Java's short-circuit evaluation of the logical AND and OR operators](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/op2.html), there would be no risk of a NullPointerException if they were combined into one `if`.

**Review by cschroedl-gov — APPROVED** [review]
Mergeworthy good stuff here! You are welcome to submit follow-up pull requests addressing the small things I was wondering about.
